### ISSUE TITLE
Issue Running MELD

### ISSUE THREAD

**Issue by nd346** [maccallumlab/meld#133]
I am facing issues running meld on the cluster, with one GPU. 
Here are the files, please guide me what is the fault with these

<img width="1512" alt="Screen Shot 2022-09-19 at 11 33 11 AM" src="https://user-images.githubusercontent.com/20399451/191055989-8f8e38a7-35ea-4d4b-8078-56f602744b9c.png">
<img width="419" alt="Screen Shot 2022-09-19 at 11 32 02 AM" src="https://user-images.githubusercontent.com/20399451/191055998-08a4f269-983f-4569-ba1f-95c41e6915bd.png">
Atribute missing is runner
Kind Regards,
Namita

**Comment by jlmaccal** [comment]
I see that you are using `launch_remd_multiplex` which is no longer part of MELD. The line numbers and functions that are giving you the error are also no longer in the code base. Let's get things setup with the latest version and go from there.

**Comment by nd346** [comment]
Hi Justin,

Sean from our institute has installed this with your help. I can request him to talk to you on the installation and as I was not the part of installation I don't know , how he did it. Is it possible for you to please direct me to the tutorial that works best with this version of the MELD, so that I can setup and run. If launch_remd_multiplex and line numbers and functions doesn't work then what will work best as I was told by Arup that is how he runs MELD.

Kind Regards,
Namita

**Comment by nd346** [comment]
From what I could find on this I followed the instructions on the below site…


Hi Justin,

This is what Sean did for installation, this is in quotes from Sean: 
"https://github.com/maccallumlab/meld

You run the below command to get the environment setup for it…

/gpfs/u/barn/MMOV/shared/etc/meld


This sets up conda, the activates the "my-meld-project” conda environment. I used to have to load CUDA, but since it gets installed in CODA, you do not need to install the system wide one. I used to set-up some paths to fftw and openmpi, but it looks like they get installed in coda as well.

So basically I get the base conda, then the meld project env setup, that’s it. "

Kind Regards,
Namita

**Comment by nd346** [comment]
Hi Justin,

Please respond, if you have received try previous message? I am sending on the details of how MELD was installed by Sean in our lab, I am pasting the message from him again:
 this is in quotes from Sean:
"
From what I could find on this I followed the instructions on the below site…

https://github.com/maccallumlab/meld

You run the below command to get the environment setup for it…

/gpfs/u/barn/MMOV/shared/etc/meld

This sets up conda, the activates the "my-meld-project” conda environment. I used to have to load CUDA, but since it gets installed in CODA, you do not need to install the system wide one. I used to set-up some paths to fftw and openmpi, but it looks like they get installed in coda as well.

So basically I get the base conda, then the meld project env setup, that’s it. "

Is it possible for you to please direct me to the tutorial that works best with this version of the MELD, so that I can setup and run. If launch_remd_multiplex and line numbers and functions doesn't work then what will work best as I was told by Arup that is how he runs MELD.

Kind Regards,
Namita

**Comment by jlmaccal** [comment]
OK, that's a reasonable way to install MELD. However, I'm 99% sure that you are not running the current version. I think that you will need to upgrade to the latest version and re-install.

**Comment by cureboon** [comment]
Hi Justin,

Sean have upgraded and reinstall the latest version, but it now has ModuleNotFoundError: No module named ‘mpi4py’. Is it possible for you to interact with Sean regarding this. I am sharing the screen shot of the error

Kind Regards,
Namita

<img width="1512" alt="Screen Shot 2022-10-27 at 12 00 33 AM" src="https://user-images.githubusercontent.com/115676448/198189097-c73820f0-84e7-4535-ae12-ef488007d756.png">

**Comment by alberto99** [comment]
Hi Namita,
There are several things going on here. The first is that once you have run the setup script once it will generate a Data directory. By default, we do not overwrite a previously written Data directory, so some of the errors you see are just saying that it cannot actually save the configuration files.

The second thing is about the mpi4py. When you run the setup script it is working on serial, so it should not be a problem despite the warning. Once you have your configuration files in the Data directory and you submit the MELD simulation (with however many replicas you have decided), then it will need to parallelize and mpi4py should be available. I don’t know if you are using conda environments or virtualenvironments, but you should do pip install mpi4py.
Best,
Alberto

From: cureboon ***@***.***>
Date: Thursday, October 27, 2022 at 12:08 AM
To: maccallumlab/meld ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [maccallumlab/meld] Issue Running MELD (Issue #133)

Hi Justin,

Sean have upgraded and reinstall the latest version, but it now has ModuleNotFoundError: No module named ‘mpi4py’. Is it possible for you to interact with Sean regarding this. I am sharing the screen shot of the error

Kind Regards,
Namita

[Screen Shot 2022-10-27 at 12 00 33 AM]<https://user-images.githubusercontent.com/115676448/198189097-c73820f0-84e7-4535-ae12-ef488007d756.png>

**Comment by thype2000** [comment]
Hi, this is Sean. I am going to work on this in more detail now to get it running for Namita. We can not run the newest version due the glib libraries. We are running this on CentOS/RedHat 7. I had to install an older version to get it running for her.

**Comment by thype2000** [comment]
I was looking to see if there is a way to run it without using slurm, I have a test system that we do not have to use slurm, but gets me a chance to look into things with slurm if I know it runs without it first.

**Comment by thype2000** [comment]
A little more info. I installed as its called on the site from "Scratch". I tried running your setup_simulation.py script and get the following error...

**Comment by thype2000** [comment]
Traceback (most recent call last):
  File "/scratch/colles2/x/meld/setup_simulation.py", line 13, in <module>
    builder = system.AmberSystemBuilder(build_options)
NameError: name 'system' is not defined

**Comment by ccccclw** [comment]
If you have openmpi installed, you can try `mpirun -n 6 launch_remd --debug`.

**Comment by jlmaccal** [comment]
@thype2000, there is nothing specific to SLURM in how MELD works.

Are you trying to run on a single node or across multiple nodes?

**Comment by cureboon** [comment]
I am trying to run it on multiple GPUs and that is what is causing the
error.

**Comment by thype2000** [comment]
I will get Namita try this thank

**Comment by nd346** [comment]
Hi Justin, 
I have already tried it with mpirun previously and I get the error that Running with mpirun gives the error
TypeError: run() takes 3 positional arguments but 4 were given
    launch.launch(console, args.debug, args.console_log)
  File "/gpfs/u/barn/MMOV/shared/miniconda3/envs/my-meld-project/lib/python3.9/site-packages/meld/remd/launch.py", line 111, in launch
    remd_runner = store.load_remd_runner().to_follower()
AttributeError: 'MultiplexReplicaExchangeRunner' object has no attribute 'to_follower'
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
srun: Job 1023 step creation temporarily disabled, retrying (Requested nodes are busy)
**or** 
AttributeError: 'MultiplexReplicaExchangeRunner' object has no attribute 'to_follower'
    launch.launch(console, args.debug, args.console_log)
  File "/gpfs/u/barn/MMOV/shared/miniconda3/envs/my-meld-project/lib/python3.9/site-packages/meld/remd/launch.py", line 109, in launch
    remd_runner = store.load_remd_runner().to_follower()
AttributeError: 'MultiplexReplicaExchangeRunner' object has no attribute 'to_follower'
    remd_runner.run(communicator, system_runner, store)
TypeError: run() takes 3 positional arguments but 4 were given
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2

Please let me know any other way to run this I tried running with multiplex on single node it is running fine. but as soon as I try with mpirun/srun it doesn't work. Please let me know what can be the issue over here.

**Comment by jlmaccal** [comment]
You have an older version that is in your path somewhere.

The multiplex commands no longer exist.

And the error about no attribute “to_follower” also indicates that somehow old scripts are being mixed in. New versions of meld have a “to_worker” method, but the script that you are running is using “to_follower”. Indeed, if you search the source code, you will not find “to_follower” anywhere, which indicates that at least part of your installation is from an old version. Somehow, you seem to have parts of two different versions mixed together.

What does “which launch_remd” give? And does that script correspond the latest version?

**Comment by cureboon** [comment]
Hi Justin,

As you know Sean,  the system administrator has installed MELD. But when I
tried checking  which launch_remd, it shows
/gpfs/u/barn/MMOV/shared/miniconda3/envs/my-meld-project/bin/launch_remd.
and in bin when I look at the launch_remd file it does not show the
version. I have requested Sean to provide the version, he might respond
here.

Kind Regards,
Namita

**Comment by thype2000** [comment]
I looked at that script and there is no version information in it. I can say it was installed on January 10, so whatever things were then is what is there now. FYI I do see some files claiming they have "to_follower" within them, but one file that has "to_worker". At this point it seems like it might best to just do a complete reinstall from scratch, unless you know something else I am unaware of.

**Comment by jlmaccal** [comment]
I can't tell exactly what is going on, but there are almost certainly a mix of scripts from two different versions.

**Comment by thype2000** [comment]
I am just going to start from the beginning and do a clean install again, then have Namita test.

**Comment by thype2000** [comment]
OK, I took the day and tried to install the latest version, as if it was the first time. It fails every time on every OS and using different versions of Miniconda. When I try to install using conda I always get glibc errors on every OS. I try to install by hand and the python script fail as well as the plugin stuff below that. What OS/version are you using to build and test this software? What version of Python? Any other version info would be appreciated as well.

**Comment by jlmaccal** [comment]
The automated scripts that build the conda package are here: https://github.com/conda-forge/meld-feedstock

We're not doing anything unusual there, so I don't know why you would be getting glibc errors. Do any of the conda packages work?

As far as installing from source goes, you didn't say what error message you were getting.

The automated scripts used to build MELD for testing are in the .github and .devtools directories. Those are built on Ubuntu-latest with several different versions of python. We're also using it on a number of research computing systems with a variety of os versions.

I understand that you are frustrated, but please be patient. The issue really has nothing to do with meld per se, but is instead the libraries and dependencies on your system. This is really challenging to debug, especially via github.

What's the specific glibc error you get with conda? And what is the specific error you get when building from source?

**Comment by thype2000** [comment]
Here is some info, if you notice anything, let me know. I am going to debug a bunch of stuff today in more detail.

When installing from Conda here is what I am getting...

RedHat/CentOS 7

glibc >= 2.17

Ubuntu 20

glibc >= 2.31

When installing from Scratch I can get everything installed but when I run python setup.py install I get...

Traceback (most recent call last):
  File "/scratch/colles2/meld/setup.py", line 2, in <module>
    from meld import __version__
  File "/scratch/colles2/meld/meld/__init__.py", line 16, in <module>
    from meld.vault import DataStore
  File "/scratch/colles2/meld/meld/vault.py", line 16, in <module>
    from meld.vault import DataStore
  File "/scratch/colles2/meld/meld/vault.py", line 16, in <module>
    from meld.system import options
  File "/scratch/colles2/meld/meld/system/options.py", line 15, in <module>
    @partial(dataclass, frozen=True)
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/colles2/miniconda3/envs/my-meld-project/lib/python3.11/dataclasses.py", line 1223, in dataclass
    return wrap(cls)
           ^^^^^^^^^
  File "/scratch/colles2/miniconda3/envs/my-meld-project/lib/python3.11/dataclasses.py", line 1213, in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/colles2/miniconda3/envs/my-meld-project/lib/python3.11/dataclasses.py", line 958, in _process_class
    cls_fields.append(_get_field(cls, name, type, kw_only))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch/colles2/miniconda3/envs/my-meld-project/lib/python3.11/dataclasses.py", line 815, in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
ValueError: mutable default <class 'openmm.unit.quantity.Quantity'> for field pressure is not allowed: use default_factory
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------

**Comment by jlmaccal** [comment]
What is the error you get when try to use conda? I think you're going to run into trouble using centos7. That's a very old operating system, released in 2014. It is being end-of-lifed next year.

The second error might be due to a change in python 3.11. I will look into fixing it.

**Comment by thype2000** [comment]
Unfortunately we are stuck with RedHat 7 for a bit longer. This is what the super computers here are still running. I know they are looking into upgrading, but I do not know when.

I did just try this on Ubuntu 22 and got the same thing, so I have to believe I am missing something. I only do basic installs of Server OS's, then add what I need.

Errors on Ubuntu 22:

  - feature:/linux-64::__glibc==2.35=0
  - meld -> __glibc[version='>=2.17']
  - meld -> cudatoolkit[version='>=11.2,<12'] -> __glibc[version='>=2.17,<3.0.a0']
  - python=3.11 -> libgcc-ng[version='>=11.2.0'] -> __glibc[version='>=2.17']

Your installed version is: 2.35

The ererors are basically exactly the same on Ubuntu 20 and RedHat 7, as stated before.

**Comment by jlmaccal** [comment]
Can you give the command you are running and the full output. The error doesn't really make sense, as I don't see how that version constraint is violated.

I did a quick google, and I didn't find anything obvious about the inability to install conda on centos 7, so I'm surprised that there are issues.

**Comment by jlmaccal** [comment]
The issue when building from source is due to a change in the dataclass library in python 3.11. It should not be too hard to fix this, but it might take a while for us to get to it.

In the meantime, if using python 3.10 is an option, it should work.

**Comment by thype2000** [comment]
OK, your first question you asked late yesterday is an easy answer...

conda create -n my-meld-project python
conda activate my-meld-project
conda install meld

All it does is try to figure out conflicts and the then fails with the errors I gave you. The issue with all of that is I can install cuda by hand without issue, which is one of the weird problems.


Your second post says to try a different version of Pythn. I have tried all three available Miniconda versions which is 3.8, 3.9, 3.10 which gets updated to 3.11 lately.

I can install all modules to do your "Scratch" install method.

conda create -n my-meld-project python
conda activate my-meld-project
conda install -c omnia ambermini 
conda install netcdf4 
conda install openmm
conda install conda-forge::cudatoolkit==11.2.*
conda install scipy
conda install -c conda-forge scikit-learn
conda install progressbar
conda install -c conda-forge eigen
conda install openmpi mpi4py

I then, as stated before, try to run python install.py python and fails with the error stated before. I have tried different cuda versions, different compilers in my path.

From my point of view this install.py file is asking for python modules that do not exist yet in the environment. 

I will crate a new Ubuntu 22 VM and use a lower level Python, do the conda installs by hand and then try again.

**Comment by thype2000** [comment]
Your last statement got me thinking and it is an easy fix for me and your website.

you have:

conda create -n my-meld-project python

This has installed python 3.11 for a while now. Just looked at my notes, never really noticed it. So doing: 

conda create -n my-meld-project python==3.10

-or-

conda create -n my-meld-project python==3.9

both work. Perfectly fine, like it should. Even on CentOS 7. I have it installed and Namita will be able to start testing today. Thank You

**Comment by nd1984** [comment]
Hi Everyone,
Sean has installed the MELD again following your instructions.
Could you please point me towards the main tutorial with step by step instruction for running meld for protein-peptide binding. Because When I use the python script to setup meld system, that is mentioned here https://github.com/maccallumlab/meld/tree/master/docs/tutorial/protein_peptide_tutorial (setup_meld.py). 
**It gives the following error**, 
"/scratch/ndube/meld/setup_binding_meld.py:134: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if (type is 'hydrophobic'):
/scratch/ndube/meld/setup_binding_meld.py:136: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if (type is 'strand'):
/scratch/ndube/meld/setup_binding_meld.py:138: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if (type is 'Hbond'):
Traceback (most recent call last):
  File "/scratch/ndube/meld/setup_binding_meld.py", line 10, in <module>
    from meld.system.restraints import LinearRamp,ConstantRamp
ImportError: cannot import name 'LinearRamp' from 'meld.system.restraints' (/gpfs/u/barn/MMOV/shared/miniconda3/envs/my-meld-project/lib/python3.9/site-packages/meld/system/restraints.py)
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them. 

**It is also not running the meld as it is producing this error**: "ModuleNotFoundError: No module named 'meld.pdb_writer'
    return pickle.load(data)
ModuleNotFoundError: No module named 'meld.pdb_writer'
MPI node 1/2 raised exception.
MPI node 2/2 raised exception.
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them." Please help me out with this issue.

Kind Regards,
Namita

**Comment by thype2000** [comment]
These files definitely exist and are readable.

**Comment by cureboon** [comment]
Could you please point me in the right direction, what can be the issue
over here?

**Comment by jlmaccal** [comment]
> These files definitely exist and are readable.

This doesn’t make sense. ‘meld.pdb_writer’ does not exist in the current repository. So, something still seems messed up with versions.

is it possible that the data directory was not deleted after running an older version of setup_meld.py? When you run that script, it creates a bunch of Python pickle files on disk that contain the details of how to run the calculation. When you later launch remd, it reads those files to run the calculation. In general, you need to use the same version of Meld for these two steps.

**Comment by nd1984** [comment]
After transferring the files to a new directory, I still gets LinearRamp error:
ImportError: cannot import name 'LinearRamp' from 'meld.system.restraints' (/gpfs/u/barn/MMOV/shared/miniconda3/envs/my-meld-project/lib/python3.9/site-packages/meld/system/restraints.py).

Could you provide the example test files and folder. Maybe I am doing something wrong here. I am attaching the script that I am running, by using python setup_binding_meld.py script to generate the data. This use to generate the data folder successfully earlier. 

Kind Regards,
Namita

**Comment by jlmaccal** [comment]
It looks like that tutorial is a bit out of date. @alberto99, could someone from your group bring it up to compatibility with the most recent version of MELD?

Namita, in the meantime, you can import `LinearRamp` from `meld.system.scalers`.

**Comment by nd1984** [comment]
Hi Justin,

I did tried this but now getting this error AttributeError: module 'meld.system' has no attribute 'ProteinMoleculeFromPdbFile'. Can someone from your group try to help as Alberto seems busy.

Kind Regards,
Namita

**Comment by nd1984** [comment]
Hi Justin and Alberto,

I am following up on this issue, Is it possible to arrange a meeting to guide me in this regard?

Kind Regards,
Namita

**Comment by nd1984** [comment]
Thanks Arup! Are you available tomorrow

**Comment by nd1984** [comment]
Tomorrow afternoon after 2 pm works for me

**Comment by arupmondal835** [comment]
Cool. let's meet tomorrow at 3 pm then.

**Comment by cureboon** [comment]
Hi Arup,

Please send the zoom link to our meeting.

Kind regards,
Namita
